R/utils.R

```diff
@@ -585,17 +585,17 @@ str_string = function(x, collapse = " ") {
 #' check_tz_to_result(42)
 #' check_tz_to_result(NULL, allow_null = FALSE)
 check_tz_to_result = function(tz, allow_null = TRUE) {
-  if (is.null(tz) && !allow_null) {
-    stop("pre-check tz: here NULL tz is not allowed")
-  }
-  if (!is.null(tz) && # null tz is fine
-      (!is_string(tz) || !tz %in% base::OlsonNames()) # otherwise must be a string of OlsenNames
-  ) {
-    stop(paste0(
-      "pre-check tz: '", tz, "' is not a valid time zone string from base::OlsonNames() or NULL"
-    ))
-  }
-  tz
+  pcase(
+    is.null(tz) && !allow_null, Err_plain("pre-check tz: here NULL tz is not allowed"),
+
+    !is.null(tz) && (!is_string(tz) || !tz %in% base::OlsonNames()), Err_plain(
+      "pre-check tz: '",
+      tz,
+      "' is not a valid time zone string from base::OlsonNames() or NULL"
+    ),
+
+    or_else = Ok(tz)
+  )
```

Comment:
@sorhawell I had made this change because `replace_time_zone()` returns a nested `Result` that needs to be unwrapped twice. With this change, the tests for `replace_time_zone()` fail

Reply:
point taken, I was a little tired yesterday. In rust lingo we can use `and_then` when the map-ok function also produces a result, and we just want to flatten the nested result.

https://github.com/pola-rs/r-polars/pull/417/commits/c2ae4d99aea22f0269c071d4e6af9cdaf6f4e041

I have not tested this yet, just added it

Reply:
```R
#' map an Err part of Result
#' @param x any R object
#' @param f a closure that takes the ok part as input
#' @return same R object wrapped in a Err-result
#' @noRd
map = function(x, f) {
  if (is_ok(x)) x$ok = f(x$ok)
  x
}

#' map an ok-value or pass on an err-value
#' @param x any R object
#' @param f a closure that takes the ok part as input
#' @return same R object wrapped in a Err-result
#' @keywords internal
and_then = function(x, f) {
  if (is_err(x)) {
    return(x)
  }
  guard_result(f(x$ok), msg = "in and_then(x, f): f must return a result")
}
```

Reply:
No worries, it passes the tests now